### PR TITLE
Refactor and expand tests

### PR DESCRIPTION
**Pull request by jordandeklerk**
This significantly refactors and expands the test suite for arviz-stats. The goal here was to test core functionality and relevant edge cases for areas that either lacked test coverage or had very minimal exposure. I also moved all `loo` based testing into its own testing directory so we could expand on those tests and separate the modules out. 

While this PR primarily focuses on robustifying the test suite, there were several changes to source code as well along the way. Some of these are (hopefully) small improvements, while others came up from some edge-case tests.

### Source Changes

- **[`src/arviz_stats/accessors.py:276`](https://github.com/arviz-devs/arviz-stats/blob/improve-tests/src/arviz_stats/accessors.py#L276)**
   - Fixed typo in error message: changed `self._obs.name` to `self._obj.name`

- **`src/arviz_stats/numba/intervals.py`**
  - Fixed missing `return` statement in [`quantile()` function (line 31)](https://github.com/arviz-devs/arviz-stats/blob/improve-tests/src/arviz_stats/numba/intervals.py#L31)
  - Fixed in-place assignment in [`_quantile()` (line 21)](https://github.com/arviz-devs/arviz-stats/blob/improve-tests/src/arviz_stats/numba/intervals.py#L21): changed `result = ...` to `result[:] = ...`

- **`src/arviz_stats/numba/diagnostics.py`**
  - Fixed DataTree group access in [`rhat()` (line 121)](https://github.com/arviz-devs/arviz-stats/blob/improve-tests/src/arviz_stats/numba/diagnostics.py#L121): changed `ds[group]` to `ds[group].dataset`
  - Fixed DataTree group access in [`ess()` (line 286)](https://github.com/arviz-devs/arviz-stats/blob/improve-tests/src/arviz_stats/numba/diagnostics.py#L286): changed `ds[group]` to `ds[group].dataset`

- **[`src/arviz_stats/sampling_diagnostics.py:388`](https://github.com/arviz-devs/arviz-stats/blob/improve-tests/src/arviz_stats/sampling_diagnostics.py#L388)**
  - Fixed recursive call: changed `rhat` to `rhat_nested` in `data.map()` call

- **[`src/arviz_stats/base/dataarray.py:325-328`](https://github.com/arviz-devs/arviz-stats/blob/improve-tests/src/arviz_stats/base/dataarray.py#L325-L328)**
  - Added defensive fallback in thinning factor calculation to ensure `factor >= 1`
  - Prevents issues when ESS exceeds n_samples (probably unlikely in practice with real data, but is possible with synthetic data)

- **`src/arviz_stats/survival.py`** ([lines 25-27](https://github.com/arviz-devs/arviz-stats/blob/improve-tests/src/arviz_stats/survival.py#L25-L27), [lines 106-108](https://github.com/arviz-devs/arviz-stats/blob/improve-tests/src/arviz_stats/survival.py#L106-L108))
  - Improved `var_names` parameter in `kaplan_meier()` and `generate_survival_curves()`
  - Now accepts both `str` and `list[str]` and is not listed as optional anymore

- **`src/arviz_stats/base/dataarray.py`** ([line 115](https://github.com/arviz-devs/arviz-stats/blob/improve-tests/src/arviz_stats/base/dataarray.py#L115), [lines 125-129](https://github.com/arviz-devs/arviz-stats/blob/improve-tests/src/arviz_stats/base/dataarray.py#L125-L129))
  - Added `method` parameter to `rhat_nested()` with default `"rank"`

- **[`src/arviz_stats/sampling_diagnostics.py`](https://github.com/arviz-devs/arviz-stats/blob/improve-tests/src/arviz_stats/sampling_diagnostics.py#L378-L416)** (lines 378-416)
  - Propagated `method` parameter throughout all `rhat_nested()` calls
  - Ensures consistent behavior across DataArray, Dataset, and DataTree interfaces

- **[`src/arviz_stats/numba/diagnostics.py:69`](https://github.com/arviz-devs/arviz-stats/blob/improve-tests/src/arviz_stats/numba/diagnostics.py#L69)**
  - Switched from `raw_rearrange` to `rearrange` (was getting an error for raw_rearrange)

I think all fixes should be backward compatible. I know this is quite a large PR. Things escalated quickly to say the least 😄. 

---
Resolves (first pass at least): https://github.com/arviz-devs/arviz-stats/issues/218

----
📚 Documentation preview 📚: https://arviz-stats--220.org.readthedocs.build/en/220/

### CODE REVIEW

File: tests/base/test_array.py [hunk 1]
```diff
@@ -0,0 +1,467 @@
...
+
+
+@pytest.fixture
+def array_stats():
+    return BaseArray()
+
+
+class TestHelperFunctions:
+    def test_process_chain_none_with_none(self):
+        ary = np.random.randn(100, 50)
```
Comment (OriolAbril) [inline tests/base/test_array.py]: these should switch to using the newer numpy random generators. Using `RandomState` instead of `Generator` is discouraged, calling `numpy.random.function` directly even more so. Ref: https://numpy.org/doc/stable/reference/random/legacy.html#functions-in-numpy-random

EDIT: in this particular function and a few more below where we basically care about the shape only we are probably better off using `np.empty` though.

File: tests/base/test_array.py [hunk 2]
```diff
@@ -0,0 +1,467 @@
...
+        ary = rng.vonmises(0, 2, size=1000)
+        result = array_stats.hdi(ary, prob=0.9, circular=True)
+        assert result.shape == (2,)
+
+    @pytest.mark.parametrize("axis", [0, 1, -1])
+    def test_hdi_axis(self, array_stats, axis):
+        rng = np.random.default_rng(42)
+        ary = rng.normal(size=(10, 20, 30))
+        result = array_stats.hdi(ary, prob=0.9, axis=axis)
+        assert result.shape[-1] == 2
```
Comment (OriolAbril) [inline tests/base/test_array.py]: I would also check the reduced dim is the correct one

File: tests/base/test_array.py [hunk 3]
```diff
@@ -0,0 +1,467 @@
...
+        assert result.shape == ()
+        assert result > 0
+
+    def test_mcse_invalid_method(self, array_stats):
+        ary = np.random.randn(4, 100)
+        with pytest.raises(ValueError, match="must be one of"):
+            array_stats.mcse(ary, method="invalid")
+
+
+class TestParetoFunctions:
```
Comment (OriolAbril) [inline tests/base/test_array.py]: I would also add tests for ndarray input, with both integer and list/array input for `axis` argument (with the exception of pareto_min_ss which takes chain_axis and draw_axis so will probably look closer to the cases above)

File: tests/base/test_array.py [hunk 4]
```diff
@@ -0,0 +1,467 @@
...
+        result = array_stats.compute_ranks(x, relative=True)
+        assert np.all(result >= 0)
+        assert np.all(result <= 1)
+
+    @pytest.mark.parametrize("axis", [0, 1, -1])
+    def test_compute_ranks_axis(self, array_stats, axis):
+        rng = np.random.default_rng(42)
+        ary = rng.normal(size=(10, 20, 30))
+        result = array_stats.compute_ranks(ary, axis=axis)
+        assert result.ndim == ary.ndim
```
Comment (OriolAbril) [inline tests/base/test_array.py]: given the shape is always preserved independently of what is passed as `axis` I would check the maximum value within the array which will depend on the length of the dimension indicated by `axis`

File: tests/base/test_array.py [hunk 5]
```diff
@@ -0,0 +1,467 @@
...
+
+
+class TestBinning:
+    def test_get_bins_basic(self, array_stats):
+        rng = np.random.default_rng(42)
+        ary = rng.normal(size=(100,))
+        bins = array_stats.get_bins(ary)
+        assert bins.ndim == 1
+        assert len(bins) > 0
```
Comment (OriolAbril) [inline tests/base/test_array.py]: given the test is seeded maybe we should check the length of `bins` and add an assert against that?

I also forgot the combination of algorightms/rules we use as "arviz auto method" but it might also be a good idea to check bins[0] and [-1] against minimum and max of `ary`. I think the checks will be min=bins[0] and then either max=bins[-1] or max<=bins[-1]

File: tests/base/test_array.py [hunk 6]
```diff
@@ -0,0 +1,467 @@
...
+class TestKDE:
+    def test_kde_basic(self, array_stats):
+        rng = np.random.default_rng(42)
+        ary = rng.normal(size=(100,))
+        grid, pdf, bw = array_stats.kde(ary)
+        assert len(grid) == 512
+        assert len(pdf) == 512
+        assert bw > 0
+
+    @pytest.mark.parametrize("grid_len", [128, 256, 512, 1024])
```
Comment (OriolAbril) [inline tests/base/test_array.py]: I don't think this will actually take very long, but given the test nature, I think the first two are enough to check what we care about

File: tests/base/test_array.py [hunk 7]
```diff
@@ -0,0 +1,467 @@
...
+        assert len(pdf) == grid_len
+
+    def test_kde_circular(self, array_stats):
+        rng = np.random.default_rng(42)
+        ary = rng.vonmises(0, 2, size=100)
+        grid, pdf, bw = array_stats.kde(ary, circular=True)
+        assert len(grid) == 512
+        assert len(pdf) == 512
+        assert bw > 0
+
```
Comment (OriolAbril) [inline tests/base/test_array.py]: here I would also add a test for behaviour of `axis`

**Review by OriolAbril — COMMENTED** [review]
I have started going over the tests I will continue as it will take a while. So far the most recurring comment is adding tests of `axis` behaviour in `array` functions. The xarray related tests are definitely passing the argument to the array function, so it might not have any effect on actual coverage if looked at from a global perspective.

However, there are already many functions with tests for `axis` and given we have a `minimal` test job that runs tests on an env without xarray nor arviz-base I think it would be nice to try and get the global coverage of the library and the coverage of minimal test job (counted with respect to things inside `base/*` except the dataarray file).

Second common comment is try to avoid `np.random.function` for new code and use the `default_rng()` then `rng.function`

All source code related changes look good.

File: tests/base/test_array.py [hunk 8]
```diff
@@ -0,0 +1,467 @@
...
+        ary = rng.normal(size=(100,))
+        result = array_stats.get_bins(ary, bins=bins)
+        assert result.ndim == 1
+
+    def test_histogram_basic(self, array_stats):
+        rng = np.random.default_rng(42)
+        ary = rng.normal(size=(1000,))
+        counts, edges = array_stats.histogram(ary)
+        assert len(counts) == len(edges) - 1
+        assert counts.sum() == 1000
```
[the remark this answers is not in the document]
Comment (jordandeklerk) [reply tests/base/test_array.py]: I can probably handle this in another quick PR.

**Comment by jordandeklerk** [comment]
> I have started going over the tests I will continue as it will take a while. So far the most recurring comment is adding tests of `axis` behaviour in `array` functions. The xarray related tests are definitely passing the argument to the array function, so it might not have any effect on actual coverage if looked at from a global perspective.
> 
> However, there are already many functions with tests for `axis` and given we have a `minimal` test job that runs tests on an env without xarray nor arviz-base I think it would be nice to try and get the global coverage of the library and the coverage of minimal test job (counted with respect to things inside `base/*` except the dataarray file).
> 
> Second common comment is try to avoid `np.random.function` for new code and use the `default_rng()` then `rng.function`
> 
> All source code related changes look good.

Think I have all of the changes related to axes testing you mentioned, and fixed the `np.random.function` issues. Thanks for bringing that up, I always forget about this. 

Appreciate the review so far as well. Going to continue looking for places to test axis behavior where it makes sense. I feel pretty good about the `loo` tests in terms of coverage and core testing functionality. For functions or modules with no coverage (something like `survival.py`, for example), I tried to just keep it simple for now and was hoping for feedback on anything specific we wanted to test.

File: tests/loo/test_loo_expectations.py [hunk 9]
```diff
@@ -0,0 +1,216 @@
...
+    if kind == "quantile":
+        assert loo_exp_vals.shape == (2, 8)
+    else:
+        assert loo_exp_vals.shape == (8,)
+
+    assert_almost_equal(loo_exp_vals.sel({"school": "Choate"}), expected_vals, decimal=2)
+
+
+@pytest.mark.filterwarnings("ignore::UserWarning")
+@pytest.mark.filterwarnings("ignore::RuntimeWarning")
```
Comment (OriolAbril) [inline tests/loo/test_loo_expectations.py]: not the first I have seem but these are great, thanks! For legacy arviz we missed some warnings because the tests were too noisy and didn't see some deprecations until the function was removed for example

Do you know if it is possible to restrict ignores to matching errors too?

File: tests/loo/test_reloo.py [hunk 10]
```diff
@@ -0,0 +1,281 @@
...
+
+azb = importorskip("arviz_base")
+xr = importorskip("xarray")
+
+from arviz_stats import loo, reloo
+from arviz_stats.loo.wrapper import SamplingWrapper
+from arviz_stats.utils import ELPDData
+
+
+@pytest.mark.filterwarnings("ignore:Estimated shape parameter:UserWarning")
```
Comment (OriolAbril) [inline tests/loo/test_reloo.py]: this looks like the pattern to ignore warnings with matching message, TIL ❤️

**Review by OriolAbril — APPROVED** [review]
I think we can merge already as it is a clear improvement on the current situation, but I think we still need to work on that, not so much in terms of coverage number or tests themselves but more organizational level things. This was already an issue before, but I am still not sure if I make a change to the kde where should new tests go, or without running them even where tests to update will be.

File: tests/loo/test_loo_expectations.py (hunk 9 above)
Comment (jordandeklerk) [reply tests/loo/test_loo_expectations.py]: The warning messages were getting wild in the test output because we're explicitly trying to raise a lot of these warnings. So I tried to go through and make sure we're not ignoring organic warnings that surface normally.

**Comment by jordandeklerk** [comment]
> I think we can merge already as it is a clear improvement on the current situation, but I think we still need to work on that, not so much in terms of coverage number or tests themselves but more organizational level things. This was already an issue before, but I am still not sure if I make a change to the kde where should new tests go, or without running them even where tests to update will be.

Agreed, I think the organization could definitely be better. Also, the workflows you mentioned in slack for running different tests suites based on dependencies would be very nice. The tests take slightly longer to run now given the new volume of tests, but I don't think it's too crazy.
